node/src/components/network.rs

```diff
@@ -0,0 +1,610 @@
+pub(crate) use event::Event;
+pub(crate) use message::Message;
+
+/// The timeout for connection setup (including upgrades) for all inbound and outbound connections.
+// TODO - make this a config option.
+const CONNECTION_TIMEOUT: Duration = Duration::from_secs(10);
+/// Env var which, if it's defined at runtime, enables the libp2p server.
+pub(crate) const ENABLE_LIBP2P_ENV_VAR: &str = "CASPER_ENABLE_LIBP2P";
+
+/// A helper trait whose bounds represent the requirements for a payload that `Network` can
+/// work with.
+pub trait PayloadT:
+    Serialize + DeserializeOwned + Clone + Debug + Display + Send + 'static
+{
+}
+
+impl<P> PayloadT for P where
+    P: Serialize + DeserializeOwned + Clone + Debug + Display + Send + 'static
+{
+}
+
+/// A helper trait whose bounds represent the requirements for a reactor event that `Network` can
+/// work with.
+pub trait ReactorEventT<P: PayloadT>:
+    From<Event<P>> + From<NetworkAnnouncement<NodeId, P>> + Send + 'static
+{
+}
+
+impl<REv, P> ReactorEventT<P> for REv
+where
+    P: PayloadT,
+    REv: From<Event<P>> + From<NetworkAnnouncement<NodeId, P>> + Send + 'static,
+{
+}
+
+#[derive(NetworkBehaviour)]
+struct Behavior {
+    #[behaviour(ignore)]
+    our_id: NodeId,
+    ping: Ping,
+}
+
+impl NetworkBehaviourEventProcess<PingEvent> for Behavior {
+    fn inject_event(&mut self, event: PingEvent) {
+        info!("{}: {:?}", self.our_id, event);
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, DataSize)]
+enum ConnectionState {
+    Pending,
+    Connected,
+    Failed,
+}
+
+#[derive(DataSize)]
+pub(crate) struct Network<REv, P>
+where
+    REv: 'static,
+{
+    our_id: NodeId,
+    peers: HashSet<NodeId>,
+    #[data_size(skip)]
+    listening_addresses: Vec<Multiaddr>,
+    /// The addresses of known peers to be used for bootstrapping, and their connection states.
+    #[data_size(skip)]
+    known_addresses: HashMap<Multiaddr, ConnectionState>,
+    /// Channel signaling a shutdown of the network component.
+    #[data_size(skip)]
+    shutdown_sender: Option<watch::Sender<()>>,
+    server_join_handle: Option<JoinHandle<()>>,
+    _phantom_data: PhantomData<(REv, P)>,
+}
+
+impl<REv: ReactorEventT<P>, P: PayloadT> Network<REv, P> {
+    /// Creates a new small network component instance.
+    ///
+    /// If `notify` is set to `false`, no systemd notifications will be sent, regardless of
+    /// configuration.
+    #[allow(clippy::type_complexity)]
+    pub(crate) fn new(
+        event_queue: EventQueueHandle<REv>,
+        config: Config,
+        notify: bool,
+    ) -> Result<(Network<REv, P>, Effects<Event<P>>), Error> {
+        // Create a new Ed25519 keypair for this session.
+        let id_keys = Keypair::generate_ed25519();
+        let our_peer_id = PeerId::from(id_keys.public());
+        let our_id = NodeId::from(our_peer_id.clone());
+
+        // Convert the known addresses to multiaddr format and prepare the shutdown signal.
+        let known_addresses = config
```

Comment:
We should probably just give up our own formats for addresses and use the canonical format throughout? I don't see any gain from maintaining backwards compatibility with old networking configs, especially since the `MultiAddr` format is future-proof.

Reply:
I believe we've resolved this in a separate discussion now?  The plan is to maintain the small_network component while we develop the new network component, at least until we have evidence that the new component will perform adequately.